### ISSUE TITLE
hellow,why add l2_norm_layer

### ISSUE THREAD

**Issue by majinmin** [ma1112/keras-triplet-loss#1]
i run this method,but the result is not good,when i use tf.contibute.XXX.triplet_loss,the result is very good, and it without L2_Norm_Layer

**Comment by majinmin** [comment]
and the orgin_author is not add L2_Norm_Layer .i also using  #tf.keras.backend.flatten# function,but the result loss is always at margin,(without L2_Norm_Layer)

**Comment by ma1112** [comment]
Hi  @majinmin 
Thanks for showing interest in my work.
Please note that this repository is intended only to show how omoindrot's code in the https://github.com/omoindrot/tensorflow-triplet-loss repository might be used in Keras. 

When I apply L2 normalization on the embeddings in my Jupyter notebook ( i.e. `model.add(Lambda(lambda  x: K.l2_normalize(x,axis=1)))` ) I am following the architecture in the FaceNet paper ( https://arxiv.org/abs/1503.03832 ) where they do apply L2 norm on embeddings before applying triplet loss.

To help you further assist you in resolving your issues, please share more details, i.e. actual training code, including shape of your variables ( data and label), loss function, optimizer used, model architecture. 
Without that I can only guess what goes wrong, and my best guess is that you did not flatten your labels e.g. using `labels = K.flatten(labels)`.

**Comment by majinmin** [comment]
@ma1112 thank you for your answer.
i reading the paper FaceNet,yes the model add the L2_Normal_Layer，It embeds an image x into a d-dimensional Euclidean space.when i add L2_Normal_Layer,the loss is going down,my model is based on VGG16,the mnist data reading style is following:
![image](https://user-images.githubusercontent.com/19158661/92479631-5ce5fc80-f216-11ea-8371-72f3915fd0f0.png)
and part of model code include last layer(L2_Normal_Layer) : 
![image](https://user-images.githubusercontent.com/19158661/92479753-8acb4100-f216-11ea-8d55-55c6a0a6c2ab.png)
the loss is follwing your code :
![image](https://user-images.githubusercontent.com/19158661/92480021-edbcd800-f216-11ea-91f6-938252733fd0.png)
model complie and fit :
![image](https://user-images.githubusercontent.com/19158661/92480129-12b14b00-f217-11ea-931f-2e1a3805b2ec.png)
![image](https://user-images.githubusercontent.com/19158661/92480194-265cb180-f217-11ea-80b0-6e9e90890402.png)
but the result is not good:
![image](https://user-images.githubusercontent.com/19158661/92480293-49876100-f217-11ea-9c35-506060b356b4.png)
![image](https://user-images.githubusercontent.com/19158661/92480334-55732300-f217-11ea-831a-12dc782d19ed.png)

**Comment by majinmin** [comment]
this result is using TripletSemiHardLoss:
![image](https://user-images.githubusercontent.com/19158661/92481153-7ab46100-f218-11ea-819e-40771a1585d0.png)
and i did not using L2_Normal_Layer , i do not know what happend.

**Comment by majinmin** [comment]
the loss_function is following ：
![image](https://user-images.githubusercontent.com/19158661/92482217-f2cf5680-f219-11ea-803b-22c7c9c12d72.png)
![image](https://user-images.githubusercontent.com/19158661/92482288-07135380-f21a-11ea-81db-3afbeb264fb8.png)
![image](https://user-images.githubusercontent.com/19158661/92482385-26aa7c00-f21a-11ea-85ab-4e68b85bcc2f.png)
![image](https://user-images.githubusercontent.com/19158661/92482518-4772d180-f21a-11ea-9ee5-8b2b5bede8fb.png)
![image](https://user-images.githubusercontent.com/19158661/92482565-53f72a00-f21a-11ea-8efb-63b60d263ee9.png)

**Comment by ma1112** [comment]
@majinmin 
I had time today to have a look at your problem.
I advise you to either lower your batch_size when you are using the "batch_hard" version of omoindrot's triples losses ( e.g. use a batch size of 64), or try using omoindrot's "batch_all" triplet loss with the currently set batch size of 512.

Let me know whether these ideas worked for you.

**Comment by ma1112** [comment]
Due to lack of feedback on this issue, I'm closing it.
